<?xml version="1.0"?>

<!-- This XACRO initializes the plugins for ROS2 Control -->

<robot xmlns:xacro="http://www.ros.org/wiki/xacro">

    <!-- If sim_mode is NOT activated, the mecanum arduino plugin will be used. -->
    <xacro:unless value="$(arg sim_mode)">
        <ros2_control name="HardwareSystem" type="system">
            <hardware>
                <plugin>mecanumdrive_arduino/MecanumDriveArduinoHardware</plugin>
                <param name="front_left_wheel_name">front_left_joint</param>
                <param name="front_right_wheel_name">front_right_joint</param>
                <param name="back_left_wheel_name">back_left_joint</param>
                <param name="back_right_wheel_name">back_right_joint</param>
                <param name="loop_rate">30</param>
                <param name="device">/dev/serial/by-id/usb-Arduino__www.arduino.cc__Arduino_Due_Prog._Port_14238313834351619231-if00</param>
                <param name="baud_rate">57600</param>
                <param name="timeout_ms">1000</param>
                <param name="enc_counts_per_rev">2746</param>
                <param name="pid_p">20</param>
                <param name="pid_d">12</param>
                <param name="pid_i">0</param>
                <param name="pid_o">50</param>
            </hardware>
            
            <joint name="front_left_joint"> 
                <command_interface name="velocity">
                    <param name="min">-10</param> 
                    <param name="max">10</param> 
                </command_interface>
                <state_interface name="position"/>
                <state_interface name="velocity"/> 
            </joint>

            <joint name="front_right_joint"> 
                <command_interface name="velocity">
                    <param name="min">-10</param> 
                    <param name="max">10</param> 
                </command_interface>
                <state_interface name="position"/>
                <state_interface name="velocity"/> 
            </joint>

            <joint name="back_left_joint"> 
                <command_interface name="velocity">
                    <param name="min">-10</param> 
                    <param name="max">10</param> 
                </command_interface>
                <state_interface name="position"/>
                <state_interface name="velocity"/> 
            </joint>

            <joint name="back_right_joint"> 
                <command_interface name="velocity">
                    <param name="min">-10</param> 
                    <param name="max">10</param> 
                </command_interface>
                <state_interface name="position"/>
                <state_interface name="velocity"/> 
            </joint>

            <sensor name="imu">
                <state_interface name="orientation.w"/>
                <state_interface name="orientation.x"/>
                <state_interface name="orientation.y"/>
                <state_interface name="orientation.z"/>
                <state_interface name="angular_velocity.x"/>
                <state_interface name="angular_velocity.y"/>
                <state_interface name="angular_velocity.z"/>
                <state_interface name="linear_acceleration.x"/>
                <state_interface name="linear_acceleration.y"/>
                <state_interface name="linear_acceleration.z"/>
            </sensor>

        </ros2_control>

    </xacro:unless>

    <!-- If sim_mode is activated, the gazebo plugin will be used. -->
    <xacro:if value="$(arg sim_mode)">
        <ros2_control name="GazeboSystem" type="system">
            <hardware>
                <plugin>gazebo_ros2_control/GazeboSystem</plugin>
            </hardware>
            <joint name="front_left_joint"> 
                <command_interface name="velocity">
                    <param name="min">-10</param> 
                    <param name="max">10</param> 
                </command_interface>
                <state_interface name="velocity"/>
                <state_interface name="position"/> 
            </joint>

            <joint name="front_right_joint"> 
                <command_interface name="velocity">
                    <param name="min">-10</param> 
                    <param name="max">10</param> 
                </command_interface>
                <state_interface name="velocity"/>
                <state_interface name="position"/> 
            </joint>

            <joint name="back_left_joint"> 
                <command_interface name="velocity">
                    <param name="min">-10</param> 
                    <param name="max">10</param> 
                </command_interface>
                <state_interface name="velocity"/>
                <state_interface name="position"/> 
            </joint>

            <joint name="back_right_joint"> 
                <command_interface name="velocity">
                    <param name="min">-10</param> 
                    <param name="max">10</param> 
                </command_interface>
                <state_interface name="velocity"/>
                <state_interface name="position"/> 
            </joint>
        </ros2_control>

        <gazebo>
            <plugin name="gazebo_ros2_control" filename="libgazebo_ros2_control.so">
                <parameters>$(find lcl_agv_pkg)/config/my_controllers.yaml</parameters>
                <parameters>$(find lcl_agv_pkg)/config/gazebo_ros2_control_use_sim.yaml</parameters>
            </plugin>
        </gazebo>
    </xacro:if>

</robot>